storage/store/protocol_kv_store.go

```diff
@@ -133,35 +138,17 @@ func (s *ProtocolKVStore) BatchStore(lctx lockctx.Proof, rw storage.ReaderBatchW
 //   - Consider block B, whose ingestion might potentially lead to an updated KV store. For example,
 //     the KV store changes if we seal some execution results emitting specific service events.
 //   - For the key `blockID`, we use the identity of block B which _proposes_ this updated KV store.
-//   - CAUTION: The updated state requires confirmation by a QC and will only become active at the child block,
+//   - IMPORTANT: The updated state requires confirmation by a QC and will only become active at the child block,
 //     _after_ validating the QC.
 //
-// Expected errors during normal operations:
+// CAUTION: To prevent data corruption, we need to guarantee atomicity of existence-check and the subsequent
+// database write. Hence, we require the caller to acquire [storage.LockInsertBlock] and hold it until the
+// database write has been committed.
+//
+// Expected error returns during normal operations:
 // - [storage.ErrDataMismatch] if a _different_ KV store for the given stateID has already been persisted
 func (s *ProtocolKVStore) BatchIndex(lctx lockctx.Proof, rw storage.ReaderBatchWriter, blockID flow.Identifier, stateID flow.Identifier) error {
-	if !lctx.HoldsLock(storage.LockInsertBlock) {
-		return fmt.Errorf("missing required lock: %s", storage.LockInsertBlock)
-	}
-
-	existingStateID, err := s.byBlockIdCache.Get(s.db.Reader(), blockID)
-	if err == nil {
-		// no-op if the *same* stateID is already indexed for the blockID
-		if existingStateID == stateID {
-			return nil
-		}
-
-		return fmt.Errorf("kv-store snapshot with block id (%x) already exists and different (%v != %v): %w",
-			blockID[:],
-			stateID, existingStateID,
-			storage.ErrDataMismatch)
-	}
-	if !errors.Is(err, storage.ErrNotFound) { // `storage.ErrNotFound` is expected, as this indicates that no receipt is indexed yet; anything else is an exception
-		return fmt.Errorf("could not check if kv-store snapshot with block id (%x) exists: %w", blockID[:], irrecoverable.NewException(err))
-	}
-
-	return s.db.WithReaderBatchWriter(func(rw storage.ReaderBatchWriter) error {
-		return s.byBlockIdCache.PutTx(rw, blockID, stateID)
-	})
+	return s.byBlockIdCache.PutWithLockTx(lctx, rw, blockID, stateID)
```

Comment:
note for other reviewers: this check is now happening at the lowest level in [`operation.IndexProtocolKVStore`](https://github.com/onflow/flow-go/blob/af59c4c371c6b1aef1bf78ab4d1ecc664e024460/storage/operation/protocol_kv_store.go#L39-L59)